datameta/api/files.py

```diff
@@ -23,108 +23,151 @@
     """Announce a file and get a URL for uploading it."""
-    pass
-    return {}
+    db = request.dbsession
+
+    # Check authentication and raise 401 if unavailable
+    auth_user = security.revalidate_user(request)
+
+    # Extract request body fields
+    req_name = request.openapi_validated.body["name"]
+    req_checksum = request.openapi_validated.body["checksum"]
+
+    # Create the corresponding database object
+    db_file = models.File(
+            site_id           = siteid.generate(request, models.File),
+            name              = req_name,
+            checksum          = req_checksum,
+            user_id           = auth_user.id,
+            group_id          = auth_user.group_id,
+            content_uploaded  = False,
+            upload_expires    = datetime.now() + timedelta(days=1) # TODO make this configurable
+            )
 
+    # INSERT the file and flush to obtain UUID for storage_uri generation
+    db.add(db_file)
+    db.flush()
+
+    # Annotate internal storage path and obtain upload URL with request headers
+    upload_url, request_headers = storage.annotate_storage(request, db_file)
+
+    # Prepare response
+    return FileUploadResponse(
+            name              = db_file.name,
+            file_id           = db_file.site_id,
+            user_id           = db_file.user.site_id,
+            group_id          = db_file.group.site_id,
+            expires_at        = db_file.upload_expires.isoformat(),
+            url_to_upload     = upload_url,
+            request_headers   = request_headers,
+            )
 
 @view_config(
-    route_name="files_id", 
-    renderer='json', 
-    request_method="GET", 
+    route_name="files_id",
+    renderer="json",
+    request_method="GET",
     openapi=True
 )
 def get_file(request: Request) -> FileResponse:
     """Get details for a file."""
     pass
     return {}
-    
-    
+
 @view_config(
-    route_name="files_id", 
-    renderer='json', 
-    request_method="PUT", 
+    route_name="files_id",
+    renderer="json",
+    request_method="PUT",
     openapi=True
 )
 def update_file(request: Request) -> HTTPOk:
     """Update not-submitted file."""
-    pass
-    return {}
+    db = request.dbsession
+
+    # Check authentication and raise 401 if unavailable
+    auth_user = security.revalidate_user(request)
+
+    # Check if the requested resource exists and is still modifiable
+    db_file = resource.resource_by_id(db, models.File, request.matchdict['id'])
+    if db_file is None:
+        raise HTTPNotFound(json=None) # 404
+    if db_file.user_id != auth_user.id or db_file.group_id != auth_user.group_id:
+        raise HTTPForbidden(json=None) # 403
+
+    # We only allow any kind of modification until the file declared uploaded
+    if db_file.content_uploaded:
+        raise errors.get_validation_error(["Resource can no longer be modified."]) # 400
+
+    # Update properties
+    db_file.checksum  = request.openapi_validated.body['checksum']
```

Comment:
just rethinking, should we even allow checksum and name to be changed?
This would be principally a new file (and not the file that you announced with the post) and thus should also deserve a new entry in the db?

I think there is rarely any use for changing other properties than content_uploaded. If there is really something wrong with the other file properties, you probably should just delete the file entry.

However, we can also leave it like that, technically there is no problem with changing them after the announcement.